add get_confirmations() RPC and add to dashboard

dan-da opened the pull request:
Addresses #6.  (partial)

The major objective of this PR is adding the `get_confirmations` RPC.   A secondary objective is displaying the confirmations field in overview screen of the Dashboard.   Other changes are ancillary.

Server:
* `implement get_confirmations` RPC
* return confirmations field in `get_dashboard_overview_data` RPC
* add `wallet_state::get_latest_balance_height()`
* add `block_height` to tuple returned from `wallet_state::get_balance_history`

Dashboard:
* display confirmations in overview screens
* change confirmations field from `usize` to `u64`

Notes:

* The `get_confirmations()` RPC function has been simplified and made more efficient than the prototype [here](https://github.com/Neptune-Crypto/neptune-core/issues/6#issuecomment-1732668472).
* `get_latest_balance_height()` will need to be modified to take into account unsynced/abandoned utxo.  There is already #28 for this, so I think it is best to do in a follow-up PR.  Also, I'm not yet certain of the correct fix, as I did try to simply filter out by `utxo.is_abandoned` in `get_balance_history()` and that did not solve the balance/history sync issue.   suggestions welcome.
* I considered returning additional fields from `get_latest_balance_height()` and just calling it `get_latest_balance()`.  However the amount field requires an additional call+loop and since there is no present need, I decided to keep it as simple/fast as possible.
* As a first pass at optimizing I had added `block_height` field to values returned from `wallet_state::get_balance_history()`.  That fn is no longer called from `get_confirmations()` however I left the `block_height` field in place because I think it would be useful to callers of the `get_history()` RPC.  I think it would be useful to display the height in Dashboard history, but decided to leave that change for a follow-up PR.  Anyway, this note explains that diff.
* I'm a bit uncertain yet about the correct type(s) for block heights (and counts).   There is the `BlockHeight` type which the compiler treats as `i128`.  But it doesn't really make sense for a height to be signed.  Also, `i128` and `u128` both seem like overkill... how many blocks are we anticipating?    Is `u64` not enough?    The dashboard was previously using `usize` for `confirmations` field, which seems a bit loose as it could be `u32` on some platform, so I changed it to `u64` and also my reasoning is less data over the wire vs `u128`.  But I would rather just be consistent with a single type (based on `u64`?) used everywhere to represent block heights.  Perhaps the RPC and Dashboard should all be using the `BlockHeight` for `confirmations` field rather than a primitive type.  thoughts?

Testing:

* Restarted server and Dashboard, then verified that confirmations are displayed in the overview screen to the right of `synced balance`
* Verified that confirmations field is zero when a new block is mined to my wallet, and increases by 1 for each subsequent block.
* Compared the `synced balance` with top-most balance in `History` screen.  The history screen balance is higher than `synced balance`.  This is wrong but expected until #28 is fixed.

I wanted to create a unit test for `get_confirmations` RPC however it is unclear to me how to do so.  There are a few unit tests in rpc_server.rs, but afaict none of them setup a wallet with a non-zero balance which would be a requirement for my test(s).  If someone could provide such a setup fn, I would happily make use of it.

Sword-Smith commented:
I'm afraid Alan and I gave you some wrong information. The description in #6 says " The number of blocks (=confirmations) since the most recent balance update". If the last balance update is a outgoing transaction (UTXOs being spent), then any of the wallet's monitored UTXOs could have been spent, not just the last one. The monitored UTXOs are ordered by the blockheight at which they were received, and this ordering does not change. The `get_balance_height` does actually return the relevant information that `get_latest_balance_height` needs, as the returned list from `get_balance_height` can be ordered by `BlockHeight` and then the last element in that ordered list should contain the latest update. Sorry that we didn't catch that when you showed you suggestion in #6.

I would suggest:
1. Fix logic to return `BlockHeight` of latest balance update -- not latest incoming transaction as this PR's code does.
2. Write a test demonstrating correct behavior
3. Verify that the block in question belongs to the canonical chain. You can use `block_belongs_to_canonical_chain` in `archival_state.rs` for this. If it does not, then find the next-most recent balance update etc.

Sword-Smith reviewed (changes requested):
See previous comment

dan-da commented:
> Fix logic to return BlockHeight of latest balance update -- not latest incoming transaction as this PR's code does.

Done.  I think.  Please review.  ;-)    

duh, whoops!   nice catch on that.   Now that you point it out, of course the list must be traversed to find latest spent utxo.  I do wonder though if it could be possible to store the latest spent in the wallet to make this lookup o(1).  Anyway I still tried to make the fn as efficient as possible given that it is called by the Dashboard overview.  Maybe premature optimization, I dunno.

> Write a test demonstrating correct behavior, or find an existing test that sets up the relevant states and add a call to get_latest_balance_height in that.

I didn't see such a test in wallet_state.rs.  Perhaps I need to look elsewhere, or maybe you can point me to one.   Anyway, I've run out of time tonight, so will look at that next coding session.   We can keep the PR open.

> Verify that the block in question belongs to the canonical chain

haven't looked into this yet.  I will take a look and if it seems simple/clear and it works I will do it, else leave for a followup PR.

> why not just use the existing BlockHeight struct

yes!  That was my initial instinct, but I was a bit confused because the Dashboard was initially written to accept a u64 and BlockHeight::sub() returns i128.   And I wasn't sure how pedantic we are being about Block*Height* vs say Block*Count*.   ie, a strict interpretation could say that a Height is only to be used in context of a count from block 0.  Anyway, with your blessing I changed it to use BlockHeight everywhere in latest commit.

extra: I added an impl Display for Sign that just prints "-" if sign is negative. This helped me debug print a list of utxo's with amount and I figure it may be worth keeping, but happy to remove if you don't like it.

Sword-Smith commented on src/models/state/wallet/wallet_state.rs:
```diff
@@ -820,32 +820,100 @@ impl WalletState {
+    ///       also an issue for get_balance_history().
+    /// see: https://github.com/Neptune-Crypto/neptune-core/issues/28
+    pub async fn get_latest_balance_height(&self) -> Option<BlockHeight> {
+        let monitored_utxos = self.wallet_db.lock().await.monitored_utxos.clone();
+        let len = monitored_utxos.len();
+
+        if len > 0 {
+            let mut latest: BlockHeight = 0.into();
+
+            // note: would be nicer to use an iterator to find max spending_height.
+            // perhaps someday: https://github.com/Neptune-Crypto/twenty-first/issues/139
+            for i in 0..len {
+                // latest change to balance could be a spend utxo, and it could be anywhere
+                // in the monitored_utxos list, so we must check each entry in list.
+                let utxo = monitored_utxos.get(i);
+                if let Some((.., spending_height)) = utxo.spent_in_block {
+                    if spending_height > latest {
+                        latest = spending_height;
+                    }
+                }
+
+                // if latest change is an incoming utxo, we can save some work by checking only
+                // the last entry because monitored_utxos is already ordered by
+                // confirmation_height ascending
+                if i == len - 1 {
+                    if let Some((.., confirmation_height)) = utxo.confirmed_in_block {
+                        if confirmation_height > latest {
+                            latest = confirmation_height;
+                        }
+                    }
+                }
+            }
+            assert!(
+                latest > 0.into() || (len == 1 && monitored_utxos.get(0).spent_in_block.is_none())
+            );
+            Some(latest)
+        } else {
+            None
+        }
+    }
```
Very nice. This function worked when I tested it. Good job!

But we can reduce the number of branches a bit by using an `Option<Blockheight>` for the `latest` variable. And we can drop the `assert`, so the code could something like this:

```rust
pub async fn get_latest_balance_height(&self) -> Option<BlockHeight> {
        let monitored_utxos = self
            .wallet_db
            .lock()
            .await
            .monitored_utxos
            .clone();
        let len = monitored_utxos.len();

        let mut latest: Option<BlockHeight> = None;
        // note: would be nicer to use an iterator to find max spending_height.
        // perhaps someday: https://github.com/Neptune-Crypto/twenty-first/issues/139
        for i in 0..len {
            // latest change to balance could be a spent utxo, and it could be anywhere
            // in the monitored_utxos list, so we must check each entry in list.
            let utxo = monitored_utxos.get(i);
            if let Some((.., spending_height)) = utxo.spent_in_block {
                if latest.is_none() || latest.is_some_and(|x| x < spending_height)
                {
                    latest = Some(spending_height);
                }
            }

            // if latest change is an incoming utxo, we can save some work by checking only
            // the last entry because monitored_utxos is already ordered by
            // confirmation_height ascending
            if i == len - 1 {
                if let Some((.., confirmation_height)) = utxo.confirmed_in_block {
                    if latest.is_none() || latest.is_some_and(|x| x < confirmation_height)
                    {
                        latest = Some(confirmation_height);
                    }
                }
            }
        }

        latest
    }
```

dan-da replied:
I started with the modified fn from your commit and re-worked it a bit so that it finds max(confirmed_in_block) and max(spent_in_block) and filters out any utxo not on current fork.  The resulting code is a bit cleaner as well.

Sword-Smith reviewed (changes requested):
With respect to the testing, you can shoehorn tests of the `get_latest_balance_height` into the existing test `wallet_state_prune_abandoned_mutxos`. It's a bit intricate because the test previously didn't update all parts of the state as it would under real execution, so I modified the test to verify what I consider correct behavior for this new function. If you want to do test-driven development of this new function, feel free to cherry-pick this commit :)

Note that this modified test also checks for the correct behavior of property "3" in my above comment.
> Verify that the block in question belongs to the canonical chain.

Also note that this modified test requires you to move the method `get_latest_balance_height` from `wallet_state.rs` to the file containing `GlobalState`.

https://github.com/Neptune-Crypto/neptune-core/commit/2bc4858272aafc8a62238206cedeade09df1d8c2

dan-da commented:
thx!  super  helpful comments.  I'm away today, but will continue on it over the weekend.

dan-da commented on src/bin/dashboard_src/overview_screen.rs:
```diff
@@ -199,6 +200,7 @@ impl OverviewScreen {
                                 own_overview_data.syncing=resp.syncing;
                                 own_overview_data.synced_balance = Some(resp.synced_balance);
                                 own_overview_data.is_mining = resp.is_mining;
+                                own_overview_data.confirmations = resp.confirmations;
```
sure, I can make this change if you want, but I guess my thought is two-fold:

1. this could be considered premature optimization.  ie, why remove useful functionality before we actually need to? 
2. Perhaps there is a way to speed up this function so it is o(1).   more below.

Ok so my knowledge of levelDB is near zero so I'm flying a bit blind and so far I've just gone with what the code already does with regards to the `monitored_utxos` data.

The basic issue is we have a conceptual table `monitored_utxos` that is indexed and sorted by the field `confirmed_in_block`.  We have another field, `spent_in_block` without any index, but we wish to find the max value of this field.  To do so we must perform a costly loop through all items.

In a SQL database we would simply add an index to `spent_in_block` and then perform a query like `select max(spent_in_block) from monitored_utxos`.  Due to the index, this becomes very fast.

iiuc, the monitored_utxos are stored in LevelDB, and retrieved in rusty_wallet_database.rs with this:

```
        let monitored_utxos_storage = schema.new_vec::<u64, MonitoredUtxo>("monitored_utxos");
```

Ok, so my questions would be:
1. Is there a way we could add an index on `spent_in_block` in LevelDB and do something equivalent to the SQL query described above?
2. Alternatively, would it be possible to add a field to wallet state that stores the max `spent_in_block` height?   This would make the reads very fast, with a hopefully very small cost when initializing/updating state, and also very small amount of storage/mem.

Clearly I need to read up some on LevelDB, but for now I figure it is better to ask you these questions.

Sword-Smith replied:
TL;DR: It's my impression that `rusty-leveldb` is very fast as long as you don't use its iterator functionality. The only place we're doing that is in the endpoint for clearing all standings:

**Don't do this without having a very good reason**
`src/models/state/mod.rs`
```rust
pub async fn clear_all_standings_in_database(&self) {
        let mut peer_databases = self.net.peer_databases.lock().await;

        let mut dbiterator: RustyLevelDBIterator<IpAddr, PeerStanding> =
            peer_databases.peer_standings.new_iter(); // VERY SLOW!

        for (ip, _v) in dbiterator.by_ref() {
            let old_standing = peer_databases.peer_standings.get(ip);

            if old_standing.is_some() {
                peer_databases
                    .peer_standings
                    .put(ip, PeerStanding::default())
            }
        }
    }

```

**About indices**
The way I would do "indices" in level-db is by making a table that contains the relevant mapping. The `MonitoredUtxo` set is currently implemented as a list which you have to loop over to get relevant information, meaning that search times for many queries will be $O(N)$, where $N$ is the number of monitored UTXOs. My intuition and, to some extent, experience tells me that this is OK for up to probably a few thousand entries. To get faster lookup times, you could add a table that contains information about the monitored UTXOs.

We have this kind of indexing implemented for blocks through the the "block index" table where the key data type and value data type looks like this. 

```rust
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum BlockIndexKey {
    Block(Digest),       // points to block headers and file locations
    File(u32),           // points to file information
    Height(BlockHeight), // Maps from block height to list of blocks
    LastFile,            // points to last file used
    BlockTipDigest,      // points to block digest of most canonical block known
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum BlockIndexValue {
    Block(Box<BlockRecord>),
    File(FileRecord),
    Height(Vec<Digest>),
    LastFile(LastFileRecord),
    BlockTipDigest(Digest),
}
```

To get a feel for how they're used, you can have a look at the `write_block` method in `src/models/state/archival_state.rs`.

I'm thinking, though, that we can wait with implementing such a "index" table until we observe functions related to `MonitoredUtxo` being too slow?

I'm OK with including the "confirmations" number on the dashboard's overview screen like you suggest. But I would like to log how long that lookup takes, in case we do that. That lookup time could then be logged at the `Debug` level.

dan-da replied:
Ok, cool.  Thx very much for the explanation!

> I'm thinking, though, that we can wait with implementing such a "index" table until we observe functions related to MonitoredUtxo being too slow?

Sure I think that's fine.   Mainly my concern has been about usage for parties that automate thousands of transactions.  For example exchanges, traders, or merchants.   Of course by the time that is happening, there should be more dev resources available, and eventually multiple wallet options as well.

So for this PR then I will plan to leave the loop that finds max(spent_in_block).  I will add a link in the code to these comments about an index and also log time spent in the fn that looks up balance confirmations as you suggest.

I'm still curious about the process of adding an index table, and so I *might* subsequently play around with that just for learning/understanding.  But there are plenty of other things to work on also.

dan-da commented on src/rpc_server.rs:
```diff
@@ -178,6 +190,25 @@ impl RPC for NeptuneRPCServer {
+                let tip_block_header =
+                    executor::block_on(self.state.chain.light_state.get_latest_block_header());
+
+                assert!(tip_block_header.height >= latest_balance_height);
```
hmm, ok, I will try and fix property 3.  I believe you've pointed the way.

>  just the RPC thread

might still bring down the entire server process eg due to lock poisoning...?

dan-da commented:
I think this is ready for another review.  See commit messages for summary of changes.

Sword-Smith reviewed (approved):
Looks very good.

Sword-Smith commented:
> I think this is ready for another review.  See commit messages for summary of changes.

If you don't mind, could you ping me on thor <at> neptune.cash? Just want to have that channel open if that's OK.

dan-da commented on src/models/state/mod.rs:
```diff
@@ -88,6 +94,131 @@ impl GlobalState {
+        debug!("call to get_latest_balance_height() took {time_secs} seconds");
+
+        #[cfg(test)]
+        println!("call to get_latest_balance_height() took {time_secs} seconds");
```
addressed in c35022b56f683570b28230127168ff8af068f479
